cmd/iam-keys-check/main.go

```diff
@@ -0,0 +1,230 @@
+	SlackChannel    string `long:"slack-channel" description:"The Slack channel." required:"true" env:"SLACK_CHANNEL"`
+}
+
+var options Options
+var logger *zap.Logger
+
+func rowToMap(header []string, row []string) map[string]string {
+	m := map[string]string{}
+	for i, h := range header {
+		m[strings.ToLower(h)] = row[i]
+	}
+	return m
+}
+
+func getCredentialReport(iamClient *iam.IAM, tries int, pollInterval int) (*iam.GetCredentialReportOutput, error) {
+	if tries <= 0 {
+		return &iam.GetCredentialReportOutput{}, errors.New("maxmimum number of tries to get credential report reached")
+	}
+	report, err := iamClient.GetCredentialReport(&iam.GetCredentialReportInput{})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == iam.ErrCodeCredentialReportNotPresentException || aerr.Code() == iam.ErrCodeCredentialReportExpiredException {
+				_, err = iamClient.GenerateCredentialReport(&iam.GenerateCredentialReportInput{})
+				if err != nil {
+					if aerr, ok = err.(awserr.Error); ok {
+						if aerr.Code() == iam.ErrCodeLimitExceededException {
+							time.Sleep(time.Duration(pollInterval) * time.Millisecond)
+							return getCredentialReport(iamClient, tries-1, pollInterval)
+						}
+					}
+					log.Fatal(err)
+				}
+			} else if aerr.Code() == iam.ErrCodeCredentialReportNotReadyException {
+				time.Sleep(time.Duration(pollInterval) * time.Millisecond)
+				return getCredentialReport(iamClient, tries-1, pollInterval)
+			}
+		}
+		log.Fatal(err)
+	}
+	return report, nil
+}
+
+func sendAlertToSlack(usersSet stringSet, maxDays float64) error {
+
+	payload := map[string]interface{}{
+		"channel": options.SlackChannel,
+		"text":    "AWS notification",
+		"attachments": []map[string]string{
+			map[string]string{
+				"title": "Message",
+				"text":  "The following users have an active access key over " + fmt.Sprint(int(maxDays)) + " days old: " + strings.Join(usersSet.toArray(true), ", ") + "",
+			},
+		},
+	}
+
+	payloadJSON, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+
+	_, err = http.Post(options.SlackWebhookURL, "application/json", bytes.NewBuffer(payloadJSON))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func triggerCheck() {
+
+	maxDays := float64(options.MaxDays)
+
+	if maxDays == 0 {
+		log.Fatal("MaxDays must be greater than 0.")
+	}
+
+	sess := session.MustMakeSession(options.Region, options.Profile)
+	iamClient := iam.New(sess)
+
+	report, err := getCredentialReport(iamClient, 5, int(options.PollInterval))
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	reader := csv.NewReader(strings.NewReader(string(report.Content)))
+
+	header, err := reader.Read()
+	if err != nil {
+		if err != io.EOF {
+			log.Fatal(err)
+		}
+	}
+
+	usersSet := newstringSet()
+
+	for {
+
+		inRow, err := reader.Read()
+		if err != nil {
+			if err == io.EOF {
+				break
+			} else {
+				log.Fatal(err)
+			}
+		}
+
+		input := rowToMap(header, inRow)
+
+		if input["access_key_1_active"] == "true" {
```

Comment:
It would be sweet if there were some tests for users with 0 access keys, 1 access key and two access keys. Also testing the time parsing bit is working correctly.

Reply:
Users always have `2` access keys, which can be either inactive or active.  I'll add a test for time parsing.